cmd/rqlite/http/client.go

```diff
@@ -0,0 +1,186 @@
+package http
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+)
+
+// ErrNoAvailableHost indicates that the client could not find an available host to send the request to.
+var ErrNoAvailableHost = fmt.Errorf("no host available to perform the request")
+
+// ErrTooManyRedirects indicates that the client exceeded the maximum number of redirects
+var ErrTooManyRedirects = fmt.Errorf("maximum leader redirect limit exceeded")
+
+// HostChangedError indicates that the underlying request was executed on a different host
+// different from the caller anticipated
+type HostChangedError struct {
+	NewHost string
+}
+
+func (he *HostChangedError) Error() string {
+	return fmt.Sprintf("HostChanedErr: new host is '%s'", he.NewHost)
+}
+
+type ConfigFunc func(*Client)
+
+// Client is a wrapper around stock http.Client that adds "retry on another host" behaviour
+// based on the supplied configuration.
+//
+// Errors that occur while performing requests are returned as they are, and the predicate to determine whether to check
```

Comment:
Nit: shorten this line a bit so it's roughly as long as the line ending in "behaviour" above.